contracts/epochs/FlowDKG.cdc

```diff
@@ -0,0 +1,251 @@
+
+pub contract FlowDKG {
+
+    pub event BroadcastMessage(phase: UInt8, nodeID: String, content: String)
+
+    pub event FinalSubmission(nodeID: String, submission: [String])
+
+    // ================================================================================
+    // CONTRACT VARIABLES
+    // ================================================================================
+
+    // Indicates whether dkg submissions are currently being collected.
+    pub enum DKGPhase: UInt8 {
+        pub case disabled
+        pub case Phase1
+        pub case Phase2
+        pub case Phase3
+        pub case finalSubmission
+    }
+
+    // tracks the current phase of the DKG
+    access(account) var currentPhase: DKGPhase
+
+    // Indicates if a Participant resource has already been claimed by a node ID
+    // from the identity table contract
+    // Node IDs have to claim a participant once
+    // one node will use the same specific ID and Participant resource for all time
+    // `nil` or false means that there is no voting capability for the node ID
+    // true means that the participant capability has been claimed by the node
+    access(account) var nodeClaimed: {String: Bool}
+
+    /// Record of whiteboard messages keyed by phase
+    access(account) var phaseMessages: {UInt8: [Message]}
+
+    // Tracks if a node has submitted their final submission for the epoch
+    // reset every epoch
+    access(account) var nodeHasSubmitted: {String: Bool}
+
+    /// Array of unique final submissions from nodes
+    /// if a final submission is sent that matches one that already has been submitted
+    /// this array will not change at all
+    access(account) var uniqueFinalSubmissions: [[String]]
+
+    /// Tracks how many submissions have been sent
+    /// for each unique final submission
+    access(account) var uniqueFinalSubmissionCount: {Int: UInt64}
```

Comment:
I think eventually we will want to be able to determine which node submitted which key vector, since submitting an incorrect key vector is likely a slashable offense. More a note for the future though, not necessary for the first version.

Reply:
I'm still going to add it now just so we have it